### PR TITLE
[O4] Automatic subtitle conversion

### PR DESCRIPTION
**Pull request by cameel**
### Changes
This pull request adds support for subtitle codecs and automatic conversion between subtitle formats. There's also a new utility for generating sample videos for integration tests.

1) Introduced `SubtitleCodec` together with all necessary helper functions and definitions just like for video and audio codecs.
    - One important difference is that there is no translation to encoder names for subtitles. As far as I can tell encoder names match codec names so it wasn't necessary in my opinion.
2) There are no new task parameters for subtitles. Subtitles are always automatically converted to something supported by the target container and the codec is selected without user's input.
3) If conversion is impossible, the subtitle stream is considered unsupported and, depending on the value of `strip_unsupported_subtitle_streams`, either causes a validation error or is not included in the final video.
    - All conversions between `ass`, `mov_text`, `subrip` and `webvtt` are possible, but not all containers support every type. In fact `matroska` is the only one that supports more than one. There's no container that supports all of them.
    - `SUBTITLE_STREAM_WHITELIST` has been removed and is no longer used to determine whether a codec can be converted.
    - Conversion/validation is done separately for each subtitle stream - multiple streams in a single video are supported.
4) For each subtitle stream in the output file, the replace command now adds the right target codec to the ffmpeg command. This introduces some complications though. To specify a codec we must refer to a particular stream in the output file by index and the indexes can be different than in the input file due to several factors:
    - Unsupported subtitle and/or data streams may be stripped and not present in the output. This shifts the indexes of all the streams following them.
    - The command inserts the transcoded video stream at position 0. If the original file had the video stream at a different position, all the streams before it will get shifted indexes.
    - The command could theoretically get input files where N original video streams are replaced by M transcoded video streams. If M != N, the indexes get shifted. This does not happen when the command is used the way Golem does it but it's possible in a general case so I had to either block it or add code to support it. Since we have to support index shifts anyway and the change for this particular corner case required just adding/subtracting a number from all indexes, I've decided to support it.
        - To support it we need to know the number of video streams in both input files so now the command fetches metadata for both (until now it was doing it only for one) and counts the streams.
5) Added `count_streams()` and `find_stream_indexes()` helpers to `meta`.
6) It's was tricky to construct the replace command correctly given that stream numbers can shift. Bad indexes don't get detected by a test that just compares the command with expectations when your expectations are wrong too. To detect broken commands I've created an integration test for replace command with subtitle conversions that actually runs ffmpeg on a sample file.
    - To avoid committing a bunch of test videos differing only by the presence and types of subtitle streams I've created a helper function that uses ffmpeg to build a file with desired stream configuration just from video/audio/subtitle files stored in the repo. It will also allow us to easily construct input files for other integration tests in the future.

### Research
ffmpeg 4.2.1 supports the following subtitle codecs:
``` bash
ffmpeg -codecs | grep -e '[D.][E.]S[V.][I.][L.]'
```
```
 ..S... arib_caption         ARIB STD-B24 caption
 DES... ass                  ASS (Advanced SSA) subtitle (decoders: ssa ass ) (encoders: ssa ass )
 DES... dvb_subtitle         DVB subtitles (decoders: dvbsub ) (encoders: dvbsub )
 ..S... dvb_teletext         DVB teletext
 DES... dvd_subtitle         DVD subtitles (decoders: dvdsub ) (encoders: dvdsub )
 D.S... eia_608              EIA-608 closed captions (decoders: cc_dec )
 D.S... hdmv_pgs_subtitle    HDMV Presentation Graphic Stream subtitles (decoders: pgssub )
 ..S... hdmv_text_subtitle   HDMV Text subtitle
 D.S... jacosub              JACOsub subtitle
 D.S... microdvd             MicroDVD subtitle
 DES... mov_text             MOV text
 D.S... mpl2                 MPL2 subtitle
 D.S... pjs                  PJS (Phoenix Japanimation Society) subtitle
 D.S... realtext             RealText subtitle
 D.S... sami                 SAMI subtitle
 ..S... srt                  SubRip subtitle with embedded timing
 ..S... ssa                  SSA (SubStation Alpha) subtitle
 D.S... stl                  Spruce subtitle format
 DES... subrip               SubRip subtitle (decoders: srt subrip ) (encoders: srt subrip )
 D.S... subviewer            SubViewer subtitle
 D.S... subviewer1           SubViewer v1 subtitle
 DES... text                 raw UTF-8 text
 ..S... ttml                 Timed Text Markup Language
 D.S... vplayer              VPlayer subtitle
 DES... webvtt               WebVTT subtitle
 DES... xsub                 XSUB
```

There are several groups here:
- Subtitles that ffmpeg can only read (`D` and no `E`). All of them seem pretty obscure (e.g. `sami` which is theoretically supported by ASF/WMV format but is rarely used in practice). I don't even have files with those.
- Subtitles that cannot be decoded or encoded (no `D` and no `E`). I don't know if that's true for every single one of them but at least `ssa` and `srt` seem to be just aliases for other codecs. They can be used to specify subtitle codec on the command line but ffprobe then reports `ass` and `subrip`, respectively, in the file.
- Bitmap-based subtitles: `dvb_subtitle`, `dvd_subtitle`, `xsub`. They can only be converted to other bitmap-based formats. I don't have files with those so I did no mark them as supported but it should be as easy as adding them and their coversions to the right dicts.
- Text-based subtitles: `ass`, `mov_text`, `subrip`, `webvtt` and `text`.
    - `text` is just a raw text stream.
    - All the others can be converted between one another and are supported by at least one of our containers. I used [Comparison of video container formats > Subtitle/caption formats support](https://en.wikipedia.org/wiki/Comparison_of_video_container_formats#Subtitle/caption_formats_support) as a starting point and confirmed that ffmpeg can actually create all the container/subtitle combinations listed there.

### API compatibility
Since subtitles get converted automatically and there are no new task parameters, this change is completely transparent to the user. There are no changes to the public API and everything should be fully backwards-compatible.

### Dependencies
This pull request is depends on #25 and code below it and can't be easily rebased directly on `master`.

**NOTE**: I have set the base branch of this pull request to `misc-pre-o4-fixes-and-refactoring` (#25). Please remember to change the base branch back to `master` before you merge.

### CODE REVIEW

File: tests/test_integration.py [hunk 1]
```diff
@@ -1,14 +1,13 @@
...
 
 from ffmpeg_tools import codecs
 from ffmpeg_tools import commands
 from ffmpeg_tools import formats
 from ffmpeg_tools import meta
-from tests.utils import get_absolute_resource_path
+from tests.utils import get_absolute_resource_path, generate_sample_video
 
 
 class TestIntegration(TestCase):
```
Comment (nieznanysprawiciel) [inline tests/test_integration.py]: We need more integration tests and less tests that check if command is generated properly. Especially when this streams logic is so complicated.

File: tests/test_test_utils.py [hunk 2]
```diff
@@ -44,3 +50,126 @@ def test_should_not_fail_if_it_gets_a_collection_or_object(self):
         # It should not fail but the name does not have to be sensibly formatted.
         # We don't want to rely here on its exact structure.
         self.assertTrue(name.startswith("dummy_function_777_"))
+
+
+class TestTestUtilsBuildTestVideo(TestCase):
```
[the remark this answers is not in the document]
Comment (cameel) [reply tests/test_test_utils.py]: This test class does contain a test that generates a video and checks its metadata to see if it was generated correctly (`test_generate_sample_video`).

It **also** contains tests that check the command because there's a separate function doing it and they're unit tests for that function. When I was writing it I first wrote down the command I was expecting then wrote a unit test based on it and then implemented a function that generates the command. It's just TDD.

I agree that one of these tests is an integration test and the other is a unit test and mixing them in a single class is not that great. Maybe I should just put the test that generates a video in a separate class called `TestTestUtilsBuildTestVideoIntegration`? I did not want to put it in `test_integration` because it's a utility that's used by integration tests and should be already tested by the time we run them :)

File: tests/test_integration.py (hunk 1 above)
Comment (cameel) [reply tests/test_integration.py]: I agree to a degree. Please note that I actually put extra time into implementing a utility that would let me test it on actual videos (`generate_sample_video()`) because I don't even have enough sample files to do real integration tests for subtitles. All the other tests are more like unit tests. There's very little in terms of integration testing right now and if I just went with the flow that would not change a bit.

On the other hand, there will always be less integration tests than unit tests. It's just not feasible to test every corner case with integration tests because things can't be mocked and we need real videos for every test. It requires lots of of actual test data and makes the test suite very slow. We need both kinds of tests in right proportion.

I agree that currently there's too little of both types. In these recent branches I tried to change that by rigorously writing unit tests for everything I touched and also adding some integration tests where it made sense.

Also, the whole set of helpers for integration testing of ffmpeg results was left in the Golem repo when ffmpeg-tools was created. I didn't want to recreate all that here.

Comment (nieznanysprawiciel) [reply tests/test_integration.py]: I think you added great amount of tests, even more then I expected. But here I'm afraid only of this logic related to subtitles and replacing streams. We could either simplify it or make more tests.

File: ffmpeg_tools/commands.py [hunk 3]
```diff
@@ -348,17 +348,107 @@ def find_unsupported_data_streams(metadata):
...
+        # based on the file name and possibly other factors.
+        # It's best to always specify the container but ffmpeg-tools allows
+        # you not to do it for the sake of flexibility.
+        #
+        # If the target container is selected implicitly by ffmpeg, we
+        # can't really know which subtitle streams it supports.
+        # The safest bet is to say that all subtitle streams are supported.
+        # This way nothing gets stripped and ffmpeg either converts them on
+        # its own or fails, letting the user know that it can't process the video.
+        return []
```
[the remark this answers is not in the document]
Comment (cameel) [reply ffmpeg_tools/commands.py]: It's already forbidden by validations. You won't get here if you run them and `container` is `None` - which is the only situation when this special case could be triggered. It will happen only if you run it despite validations telling you not to.

It's just that the commands can work just fine with a wider range of parameters than the validations allow. For example in they allow you not to specify the container and rely on the name of the output file. I kept this case working after we switched to always specifying the container simply because it was very easy to do (we already had it working) and I wasn't 100% sure at the time that we won't ever revert to relying on names. ffmpeg-tools is a library and I think it's fine to provide a bit more generality than the client needs.

But yeah, in practice Golem (and probably your service too) does not rely on this feature and as we add more functionality it seems ever more superfluous because some of that functionality (like subtitle conversion) cannot work if we don't know the target container. I think that keeping it working by not assuming anything and just hoping that ffmpeg can convert the subtitles is consistent with letting it decide the container too and I'd leave it as is here.

After this point we should probably just remove any support for not specifying the container in commands but I didn't want to introduce such a change as a part of this pull request.

[the remark this answers is not in the document]
Comment (nieznanysprawiciel) [reply ffmpeg_tools/commands.py]: So you say that if we execution will get to subtitles validation it should fail earlier if format was None?
Ok I missed it.

Let it be as is. But maybe throw exception in validate_unsupported_subtitle_streams if format is None and write comment nearby.
Because now this check relies on assumption, that other function behaves as expected. And if format validation will change, subtitles validation will change behavior implicite.

File: ffmpeg_tools/commands.py [hunk 4]
```diff
@@ -449,6 +552,20 @@ def replace_streams_command(input_file,
...
+            data_streams_to_strip +
+            subtitle_streams_to_strip +
+            meta.find_stream_indexes(input_metadata, VALID_STREAM_TYPES[stream_type]),
+        ),
+        meta.count_streams(replacement_metadata, VALID_STREAM_TYPES[stream_type]),
+    )
+    subtitle_codec_options = [
+        [f"-codec:{index}", codec]
+        for index, codec in subtitle_codec_map.items()
+    ]
```
[the remark this answers is not in the document]
Comment (cameel) [reply ffmpeg_tools/commands.py]: It would indeed make the implementation a bit simpler. But it's not really about the API. I could make it that simple even by just converting the current code into a loop over all the input streams from the two input files (though wrapping it in a class of course has additional benefits).

The reason why it's more complex now is that I tried to keep the ffmpeg command simple. Note for example that before adding subtitles the mappings boiled down to `-map 1:v -map 0 -map -0:v` (take all streams from input `#0` except for video streams and add all video streams from input `#1`). Before we had support for subtitles and data streams it allowed me to perform this operation without having to get metadata for the streams and have ffmpeg do all the dirty work for me.

Obviously, now the command does have to get metadata so what you suggest makes sense. If we already have to get it, we can just as well use it and just loop over it. By the time I had this metadata though I already had this implemented like this so I didn't want to just immediately rewrite it again :)

Having said that, I think such a redesign would be a good thing but I'd rather limit myself to fixing bugs and defects here. Let's finish getting these pull requests through the review and doing any fixes critical to the functionality first. I'm going to fix any stuff that does not work as specified in the tasks we agreed on (and even do some minor tweaks you request as long as that's not too labor-intensive) but if there's going to be more work (like further tasks or refactoring), we should discuss that in the main e-mail thread. And unfortunately my availability in the coming months will be somewhat limited as I have work on other projects queued up (though it really depends on the terms).

As for estimation:
- Doing just a loop sounds like 1 day tops. It would be all contained in the replace function and covered by already existing tests.
- `ReplaceCommandBuilder` class - maybe two or three days, depending on how generic and well tested you'd want it to be. If not very generic then maybe it would be even just 1 day but there's some uncertainty so I don't want to promise that.
- As far as refactoring goes, having something even more general, like a `FfmpegCommand` class that lets you build any of the commands and hides the ugly details inside (like matching inputs to outputs, inserting parameters in the right places in the command (ffmpeg is pretty sensitive to that) detecting contradictions, preventing duplicates, etc.) would be a great idea in my opinion. But that's basically a full rewrite of the `commands` module so significantly more work, especially when you take testing all the corner cases into account.

[the remark this answers is not in the document]
Comment (nieznanysprawiciel) [reply ffmpeg_tools/commands.py]: Let's take minimalistic approach than. We will leave it as is.
But QA will test it precisely, because I don't know if we can trust this code.

File: tests/test_integration.py (hunk 1 above)
Comment (cameel) [reply tests/test_integration.py]: OK. I addressed simplifying the logic in a comment above (https://github.com/golemfactory/ffmpeg-tools/pull/26#discussion_r353795532).

But as a short term solution - if you have any particular test in mind I can add it or try to convert it to an integration test.

File: tests/test_test_utils.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (nieznanysprawiciel) [reply tests/test_test_utils.py]: I think we can leave it as is

File: ffmpeg_tools/commands.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (cameel) [reply ffmpeg_tools/commands.py]: OK. I'll add an exception and a comment.

I'd personally use an `assert` for something like this since this is something that's impossible to happen unless someone changes the code and introduces a bug - which should be caught in development.

[the remark this answers is not in the document]
Comment (cameel) [reply ffmpeg_tools/commands.py]: Done: 87d0fce636f3b6a9d359e45d2bcab449e3f4e00a
